src/frontend/js/pages/TeacherDashboardContractsLayout/hooks/useGenerateContractArchive.tsx

```diff
@@ -0,0 +1,76 @@
+import { useEffect, useRef, useState } from 'react';
+import useContractArchive from 'hooks/useContractArchive';
+import { Organization } from 'types/Joanie';
+import { Nullable } from 'types/utils';
+
+const CONTRACT_ARCHIVE_PULL_DELAY_MS = 1000;
+
+interface UseContractGenerateArchiveProps {
+  contractArchiveId?: Nullable<string>;
+  organizationId: Organization['id'];
+}
+
+// FIXME: it's look like check option request are sent 2 by 2!
+const useGenerateContractArchive = ({
+  organizationId,
+  contractArchiveId = null,
+}: UseContractGenerateArchiveProps) => {
+  const {
+    methods: { create: createArchive, check: checkArchiveExist },
+  } = useContractArchive();
+
+  const timeoutRef = useRef<NodeJS.Timeout>();
+  const [isContractArchiveReady, setIsContractArchiveReady] = useState(false);
+
+  // Request contract archive generation to Joanie backend
+  // then start pulling if contract's archive exist to update
+  // isContractArchiveReady state.
+  const doGenerateArchive = async () => {
+    setIsContractArchiveReady(false);
+    const newContractArchiveId = await createArchive(organizationId);
+
+    // TODO maybe handle contractArchiveId === null
+    if (newContractArchiveId === null) {
+      throw new Error('contractArchiveId should not be null');
+    }
+
+    doCheckArchiveExist(newContractArchiveId);
+    return newContractArchiveId;
+  };
+
+  const doCheckArchiveExist = async (archiveId: string) => {
+    // TODO: CORS ERROR !!! to fix
+    console.log('doCheckArchiveExist::useGenerateContractArchive');
+    const isExist = await checkArchiveExist(archiveId);
+    setIsContractArchiveReady(isExist);
+
+    clearTimeout(timeoutRef.current);
+    if (!isExist) {
+      timeoutRef.current = setTimeout(
+        () => doCheckArchiveExist(archiveId),
+        CONTRACT_ARCHIVE_PULL_DELAY_MS,
+      );
+    }
+  };
+
+  // If a contractArchiveId have been given as props
+  // start pulling if contract's archive exist on mount.
+  // Also assure that we stop pulling when unmounting
+  useEffect(() => {
+    if (contractArchiveId) {
+      doCheckArchiveExist(contractArchiveId);
+    }
+
+    return () => {
+      clearTimeout(timeoutRef.current);
+    };
+  });
```

Comment:
I feel this logic relies more about archive download not generation. I wonder if we do not exec this check into the useDownloadContractArchive ?